File: src/tools/shared/dbPortGuard.ts
import type { DraftServiceSpec } from "./specSchemas";

interface DbPortEntry {
  imagePattern: RegExp;
  containerPorts: number[];
  label: string;
}

export const DB_PORT_MAP: DbPortEntry[] = [
  { imagePattern: /^postgres(:|$)/, containerPorts: [5432], label: "postgres" },
  { imagePattern: /^mysql(:|$)/, containerPorts: [3306], label: "mysql" },
  { imagePattern: /^mariadb(:|$)/, containerPorts: [3306], label: "mariadb" },
  { imagePattern: /^mongo(:|$)/, containerPorts: [27017], label: "mongo" },
  { imagePattern: /^redis(:|$)/, containerPorts: [6379], label: "redis" },
];

export interface DbPortExposureIssue {
  service: string;
  image: string;
  containerPort: number;
  hostPort: number;
  message: string;
}

export function checkDbPortExposure(
  services: Record<string, DraftServiceSpec>,
): DbPortExposureIssue[] {
  const issues: DbPortExposureIssue[] = [];
  for (const [svcName, spec] of Object.entries(services)) {
    const entry = DB_PORT_MAP.find((e) => e.imagePattern.test(spec.image));
    if (!entry) continue;
    for (const portValue of spec.ports ?? []) {
      const parsed = parseHostAndContainerPort(portValue);
      if (!parsed) continue;
      if (entry.containerPorts.includes(parsed.hostPort)) {
        issues.push({
          service: svcName,
          image: spec.image,
          containerPort: parsed.containerPort,
          hostPort: parsed.hostPort,
          message: `database ${entry.label} default port ${parsed.hostPort} is published to host port ${parsed.hostPort}; remove the port mapping or use a non-default container port — the service is reachable from other compose services without publishing`,
        });
      }
    }
  }
  return issues;
}

function parseHostAndContainerPort(
  value: string,
): { hostPort: number; containerPort: number } | null {
  const trimmed = value.trim();
  const slash = trimmed.lastIndexOf("/");
  const body = slash >= 0 ? trimmed.slice(0, slash) : trimmed;
  const parts = body.split(":");
  if (parts.length < 2) return null;
  const hostStr = parts.length === 2 ? parts[0] : parts[parts.length - 2];
  const containerStr = parts[parts.length - 1];
  if (hostStr === undefined || containerStr === undefined) return null;
  const hostMatch = hostStr.match(/^(\d+)$/);
  const containerMatch = containerStr.match(/^(\d+)$/);
  if (!hostMatch || !containerMatch) return null;
  return { hostPort: Number(hostMatch[1]), containerPort: Number(containerMatch[1]) };
}

File: python/src/docker_agent/tools/shared/db_port_guard.py
"""Database default port exposure guard.

Parity: ``src/tools/shared/dbPortGuard.ts``.
"""

from __future__ import annotations

import re
from dataclasses import dataclass

from docker_agent.tools.shared.spec_schemas import DraftServiceSpec


@dataclass(frozen=True)
class DbPortEntry:
    image_pattern: re.Pattern[str]
    container_ports: list[int]
    label: str


DB_PORT_MAP: list[DbPortEntry] = [
    DbPortEntry(re.compile(r"^postgres(:|$)"), [5432], "postgres"),
    DbPortEntry(re.compile(r"^mysql(:|$)"), [3306], "mysql"),
    DbPortEntry(re.compile(r"^mariadb(:|$)"), [3306], "mariadb"),
    DbPortEntry(re.compile(r"^mongo(:|$)"), [27017], "mongo"),
    DbPortEntry(re.compile(r"^redis(:|$)"), [6379], "redis"),
]


@dataclass
class DbPortExposureIssue:
    service: str
    image: str
    container_port: int
    host_port: int
    message: str


def _parse_host_and_container_port(value: str) -> dict[str, int] | None:
    trimmed = value.strip()
    slash = trimmed.rfind("/")
    body = trimmed[:slash] if slash >= 0 else trimmed
    parts = body.split(":")
    if len(parts) < 2:
        return None
    host_str = parts[0] if len(parts) == 2 else parts[-2]
    container_str = parts[-1]
    host_match = re.match(r"^(\d+)$", host_str)
    container_match = re.match(r"^(\d+)$", container_str)
    if not host_match or not container_match:
        return None
    return {
        "host_port": int(host_match.group(1)),
        "container_port": int(container_match.group(1)),
    }


def check_db_port_exposure(
    services: dict[str, DraftServiceSpec],
) -> list[DbPortExposureIssue]:
    """Block publishing DB default container port to the same host port."""
    issues: list[DbPortExposureIssue] = []
    for svc_name, spec in services.items():
        entry = next(
            (e for e in DB_PORT_MAP if e.image_pattern.search(spec.image)),
            None,
        )
        if entry is None:
            continue
        for port_value in spec.ports or []:
            parsed = _parse_host_and_container_port(port_value)
            if parsed is None:
                continue
            if parsed["host_port"] in entry.container_ports:
                issues.append(
                    DbPortExposureIssue(
                        service=svc_name,
                        image=spec.image,
                        container_port=parsed["container_port"],
                        host_port=parsed["host_port"],
                        message=(
                            f"database {entry.label} default port {parsed['host_port']} "
                            f"is published to host port {parsed['host_port']}; "
                            "remove the port mapping or use a non-default container port "
                            "— the service is reachable from other compose services "
                            "without publishing"
                        ),
                    )
                )
    return issues


__all__ = [
    "DB_PORT_MAP",
    "DbPortEntry",
    "DbPortExposureIssue",
    "check_db_port_exposure",
]